README.md
# ansible-modules-consul-acl

Ansible modules for the [Consul ACL system](https://www.consul.io/docs/agent/acl-system.html):

-   `consul_acl_policy`
-   `consul_acl_token`

## Installation

Install using pip:

```
pip install ansible-modules-consul-acl
```

The modules have no external dependencies except Ansible.

## Usage

The documentation for each module is mostly complete - use `ansible-doc` to view it.

### Example

<!--prettier-ignore-->
```yaml
- name: Create ACL policy
  consul_acl_policy:
    name: example
    # Rules specified as an HCL string
    rules: |
      service "example" {
        policy = "write"
      }
    state: present
    url: https://localhost:8500
    token: a22c5e4f-0f48-4907-82db-843c6baf75be # Requires acl:write
  register: consul_acl_policy

- name: Create ACL token
  consul_acl_token:
    description: Example token
    # Policies specified as a list of PolicyLink objects: https://www.consul.io/api/acl/tokens.html#policies
    policies:
      - id: "{{ consul_acl_policy.id }}"
    local: true
    state: present
    url: https://localhost:8500
    token: a22c5e4f-0f48-4907-82db-843c6baf75be # Requires acl:write
  register: consul_acl_token
```

### Environment variables

Some of the environment variables for the [Consul CLI](https://www.consul.io/docs/commands/index.html#environment-variables) will be used if they are defined:

-   `CONSUL_HTTP_ADDR` for the `url` parameter. Prefix with `https://` instead of setting `CONSUL_HTTP_SSL=true`
-   `CONSUL_HTTP_TOKEN` for the `token` parameter
-   `CONSUL_CLIENT_CERT` for the `client_cert` parameter
-   `CONSUL_CLIENT_KEY` for the `client_key` parameter

## Testing locally

To run the functional tests, set the following environment variables
from the project root directory:

```sh
export ANSIBLE_LIBRARY="$PWD/ansible/modules/consul_acl"
export ANSIBLE_MODULE_UTILS="$PWD/ansible/module_utils"
```

Then run the test playbooks in a Python environment without `ansible-modules-consul-acl` installed.


CHANGELOG.md
# Changelog

## [Unreleased]

## [0.4.0] - 2020-05-26

### Added

-   Allow setting pre-defined secret_id. ([#3](https://github.com/jsundh/ansible-modules-consul-acl/pull/3))

### Fixed

-   Fix `secret_id` not being returned by the module in certain cases.

## [0.3.0] - 2020-01-24

### Added

-   Add ability to create tokens with a specific accessor ID.

### Changed

-   Extract common API code into a module utility.

### Fixed

-   Fix adding policies to bare tokens.

## [0.2.0] - 2019-02-24

### Changed

-   Improve error handling for API request errors.

### Fixed

-   Remove duplicate key in documentation.
-   Fix token comparison with no current policies.

## [0.1.0] - 2019-02-18

### Added

-   Repository documentation.

### Fixed

-   Fix documentation in modules.

## 0.0.1 - 2019-02-17

### Added

-   `consul_acl_policy` module.
-   `consul_acl_token` module.

[unreleased]: https://github.com/jsundh/ansible-modules-consul-acl/compare/0.4.0...HEAD
[0.4.0]: https://github.com/jsundh/ansible-modules-consul-acl/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/jsundh/ansible-modules-consul-acl/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/jsundh/ansible-modules-consul-acl/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/jsundh/ansible-modules-consul-acl/compare/0.0.1...0.1.0


ansible/module_utils/consul_acl.py
import json
from ansible.module_utils.common.dict_transformations import (
    camel_dict_to_snake_dict,
    snake_dict_to_camel_dict,
)
from ansible.module_utils.urls import fetch_url


class ConsulApi(object):
    def __init__(self, module):
        self.module = module
        self.url = module.params["url"]
        self.token = module.params["token"]

    def make_request(self, endpoint, method, data=None, ignore_error=None):
        if data is not None:
            data = json.dumps(snake_dict_to_camel_dict(data))

        endpoint_url = self.url + "/v1/" + endpoint
        headers = {"Content-Type": "application/json", "X-Consul-Token": self.token}

        response, info = fetch_url(
            self.module, endpoint_url, data=data, headers=headers, method=method
        )

        status_code = info["status"]
        if ignore_error == status_code:
            return None
        elif status_code >= 400:
            self.module.fail_json(
                msg="API request failed",
                endpoint=endpoint_url,
                method=method,
                status=status_code,
                response=info["body"],
            )
        elif response is None:
            self.module.fail_json(**info)

        body = json.loads(response.read())
        if type(body) is bool:
            return body
        elif type(body) is list:
            return [camel_dict_to_snake_dict(e) for e in body]
        else:
            return camel_dict_to_snake_dict(body)


ansible/modules/consul_acl/consul_acl_policy.py
#!/usr/bin/python

# Copyright: (c) 2019, Jakob Sundh <jsundh@users.noreply.github.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

ANSIBLE_METADATA = {
    "metadata_version": "1.1",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: consul_acl_policy

short_description: Manage Consul ACL policies

version_added: "N/A"

description:
  - "Create, update and delete policies for the new Consul ACL system."
  - "Official API documentation: https://www.consul.io/api/acl/policies.html"

options:
  id:
    description:
      - The policy ID.
      - If not provided, the name will be used to match an existing policy.
    type: str
  name:
    description:
      - Name of the ACL policy.
      - Required when C(state=present).
    type: str
  description:
    description:
      - Free form human readable description of the policy.
    type: str
    default: ""
  rules:
    description:
      - "Policy rules following the HCL rule specification: https://www.consul.io/docs/agent/acl-rules.html#rule-specification."
      - Required when C(state=present).
    type: str
  state:
    description:
        - If C(present), the policy will be created or updated.
        - If C(absent), the policy will be removed if it exists.
    type: str
    default: present
    choices: [ present, absent ]
  url:
    description:
      - URL for Consul.
    type: str
    required: true
    default: the environment variable CONSUL_HTTP_ADDR
  token:
    description:
      - ACL token to use for requests.
    type: str
    required: true
    default: the environment variable CONSUL_HTTP_TOKEN
  validate_certs:
    description:
      - If C(no), SSL certificates will not be validated.
      - This should only set to C(no) used on personally controlled sites using self-signed certificates.
    type: bool
    default: yes
  client_cert:
    description:
      - PEM formatted certificate chain file to be used for SSL client authentication.
      - This file can also include the key as well, and if the key is included, I(client_key) is not required
    type: path
    default: the environment variable CONSUL_CLIENT_CERT
  client_key:
    description:
      - PEM formatted file that contains your private key to be used for SSL client authentication.
      - If I(client_cert) contains both the certificate and key, this option is not required.
    type: path
    default: the environment variable CONSUL_CLIENT_KEY

author:
    - Jakob Sundh (@jsundh)
"""

EXAMPLES = """
- name: Create policy
  consul_acl_policy:
    name: example
    description: An example policy
    rules: |
      kv_prefix "example" {
        policy = "write"
      }
    state: present
    url: http://localhost:8500
    token: "consul-management-token"
  register: consul_acl_policy

- name: Update policy by ID
  consul_acl_policy:
    name: example
    id: "{{ consul_acl_policy.id }}"
    description: An updated example policy
    rules: |
      kv_prefix "example" {
        policy = "read"
      }
    state: present
    url: http://localhost:8500
    token: "consul-management-token"

- name: Remove policy by name
  consul_acl_policy:
    name: example
    state: absent
    url: http://localhost:8500
    token: "consul-management-token"
"""

RETURN = """
operation:
  description: Operation performed - either 'create', 'update', 'delete' or 'none'
  returned: always
  type: str
  sample: create
id:
  description: The ID of the policy
  returned: when operation != 'none'
  type: str
  sample: 6a524d8b-b8d5-4e81-82b3-625aeefb40f6
hash:
  description: The policy content hash
  returned: when state == 'present'
  type: str
  sample: RV0EhBF/LkDrzjrI+4AMm1MqlX6X/J2JgW4S9uBkBu0=
"""

import os
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.consul_acl import ConsulApi


class ConsulAclPolicy(object):
    def __init__(self, module, api):
        self.module = module
        self.api = api

    def find_existing_policy(self, name):
        policies = self.api.make_request("acl/policies", "GET")
        for policy in policies:
            if policy["name"] == name:
                return policy["id"]

        return None

    def create_policy(self, **kwargs):
        created = self.api.make_request("acl/policy", "PUT", data=kwargs)

        return dict(changed=True, operation="create", **created)

    def update_policy(self, policy_id, **kwargs):
        current = self.api.make_request("acl/policy/" + policy_id, "GET")

        if not has_policy_changed(current, kwargs):
            return dict(changed=False, operation="none", **current)

        updated = self.api.make_request("acl/policy/" + policy_id, "PUT", data=kwargs)
        return dict(changed=True, operation="update", **updated)

    def delete_policy(self, policy_id):
        succeeded = self.api.make_request("acl/policy/" + policy_id, "DELETE")
        if not succeeded:
            self.module.fail_json(msg="Policy deletion failed")

        return dict(changed=True, operation="delete", id=policy_id)


def has_policy_changed(current, updated):
    for k, v in updated.items():
        # Empty list default for datacenters
        if current.get(k, []) != v:
            return True

    return False


def main():
    argument_spec = dict(
        id=dict(type="str", default=None),
        name=dict(type="str", default=None),
        rules=dict(type="str", default=None),
        description=dict(type="str", default=""),
        datacenters=dict(type="list", elements="str", default=None),
        state=dict(type="str", choices=["present", "absent"], default="present"),
        url=dict(type="str", default=os.environ.get("CONSUL_HTTP_ADDR")),
        token=dict(type="str", default=os.environ.get("CONSUL_HTTP_TOKEN")),
        validate_certs=dict(type="bool", default=True),
        client_cert=dict(type="path", default=os.environ.get("CONSUL_CLIENT_CERT")),
        client_key=dict(type="path", default=os.environ.get("CONSUL_CLIENT_KEY")),
    )

    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)

    name = module.params["name"]
    policy_id = module.params["id"]
    rules = module.params["rules"]
    description = module.params["description"]
    datacenters = module.params["datacenters"] or []
    state = module.params["state"]
    url = module.params["url"]
    token = module.params["token"]

    if not url:
        module.fail_json(msg="url must be set")
    if not token:
        module.fail_json(msg="token must be set")
    if state == "present" and not name:
        module.fail_json(msg="name must be set when state == 'present'")
    if state == "present" and not rules:
        module.fail_json(msg="rules must be set when state == 'present'")

    api = ConsulApi(module)
    consul_acl = ConsulAclPolicy(module, api)

    if not policy_id:
        policy_id = consul_acl.find_existing_policy(name)

    kwargs = dict(
        name=name, rules=rules, description=description, datacenters=datacenters
    )
    if state == "present" and policy_id:
        result = consul_acl.update_policy(policy_id, **kwargs)
    elif state == "present":
        result = consul_acl.create_policy(**kwargs)
    elif state == "absent" and policy_id:
        result = consul_acl.delete_policy(policy_id)
    else:
        result = dict(changed=False, operation="none")

    module.exit_json(**result)


if __name__ == "__main__":
    main()


ansible/modules/consul_acl/consul_acl_token.py
#!/usr/bin/python

# Copyright: (c) 2019, Jakob Sundh <jsundh@users.noreply.github.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

ANSIBLE_METADATA = {
    "metadata_version": "1.1",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: consul_acl_token

short_description: Manage Consul ACL tokens

version_added: "N/A"

description:
  - "Create, update and delete tokens for the new Consul ACL system."
  - "Official API documentation: https://www.consul.io/api/acl/tokens.html"

options:
  accessor_id:
    description:
     - The accessor ID of either an existing token to update or a non-existing to create.
     - If not specified, Consul will generate an UUID when creating a token.
    type: str
    required: false
  secret_id:
    description:
     - The secret ID of the token.
     - If not specified, Consul will generate an UUID when creating a token.
    type: str
    required: false
  description:
    description:
      - Free form human readable description of the token.
    type: str
    default: ""
  policies:
    description:
      - The list of policies that should be applied to the token.
      - Each element should be an object with an "id" and/or "name" field for an existing policy.
      - Required when C(state=present).
    type: list
    element: dict
    required: false
  local:
    description:
      - If true, the token will be local to the current datacenter.
    type: bool
    default: false
  match_description:
    description:
      - If true, the module will attempt to match an existing token based on description by listing all tokens.
      - Allows matching existing tokens without saving the accessor ID.
      - Only use this if you are sure that the token descriptions are sufficiently unique.
    type: bool
    default: false
  state:
    description:
        - If C(present), a token will be created or updated.
        - If C(absent), the token will be removed if it exists.
        - If C(cloned), the given token accessor ID will be cloned.
    type: str
    default: present
    choices: [ present, absent, cloned ]
  url:
    description:
      - URL for Consul.
    type: str
    required: true
    default: the environment variable CONSUL_HTTP_ADDR
  token:
    description:
      - ACL token to use for requests.
    type: str
    required: true
    default: the environment variable CONSUL_HTTP_TOKEN
  validate_certs:
    description:
      - If C(no), SSL certificates will not be validated.
      - This should only set to C(no) used on personally controlled sites using self-signed certificates.
    type: bool
    default: yes
  client_cert:
    description:
      - PEM formatted certificate chain file to be used for SSL client authentication.
      - This file can also include the key as well, and if the key is included, I(client_key) is not required
    type: path
    default: the environment variable CONSUL_CLIENT_CERT
  client_key:
    description:
      - PEM formatted file that contains your private key to be used for SSL client authentication.
      - If I(client_cert) contains both the certificate and key, this option is not required.
    type: path
    default: the environment variable CONSUL_CLIENT_KEY

author:
    - Jakob Sundh (@jsundh)
"""

EXAMPLES = """
"""

RETURN = """
operation:
  description: Operation performed - either 'create', 'update', 'delete' or 'none'
  returned: always
  type: str
  sample: create
accessor_id:
  description: The accessor ID of the token
  returned: when state != 'absent'
  type: str
  sample: 6a524d8b-b8d5-4e81-82b3-625aeefb40f6
secret_id:
  description: The secret ID of the token
  returned: when state != 'absent'
  type: str
  sample: 45a3bd52-07c7-47a4-52fd-0745e0cfe967
hash:
  description: The token content hash
  returned: when state == 'present'
  type: str
  sample: RV0EhBF/LkDrzjrI+4AMm1MqlX6X/J2JgW4S9uBkBu0=
description:
  description: Free form human readable description of the token
  returned: when state != 'absent'
  type: str
  sample: An example token
policies:
  description: The list of policy links connected to the token
  returned: when state != 'absent'
  type: list
"""

import os
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.consul_acl import ConsulApi
from itertools import chain


class ConsulAclToken(object):
    def __init__(self, module, api):
        self.module = module
        self.api = api

    def get_token(self, accessor_id):
        return self.api.make_request(
            "acl/token/" + accessor_id, "GET", ignore_error=403
        )

    def find_token(self, description, policies):
        policy_filter = ""
        for link in policies:
            if "id" in link:
                policy_filter = "?policy=" + link["id"]
                break

        tokens = self.api.make_request("acl/tokens" + policy_filter, "GET")
        for token in tokens:
            if token["description"] == description:
                return self.get_token(token["accessor_id"])

        return None

    def create_token(self, **kwargs):
        created = self.api.make_request("acl/token", "PUT", kwargs)

        return dict(changed=True, operation="create", **created)

    def update_token(self, current, **kwargs):
        if current.get("policies") is None:
            current["policies"] = []

        if not has_token_changed(current, kwargs):
            return dict(changed=False, operation="none", **current)

        accessor_id = current["accessor_id"]
        updated = self.api.make_request("acl/token/" + accessor_id, "PUT", data=kwargs)

        return dict(changed=True, operation="update", **updated)

    def delete_token(self, accessor_id):
        succeeded = self.api.make_request("acl/token/" + accessor_id, "DELETE")
        if not succeeded:
            self.module.fail_json(msg="Token deletion failed")

        return dict(changed=True, operation="delete", accessor_id=accessor_id)

    def clone_token(self, accessor_id):
        cloned = self.api.make_request("acl/token/{}/clone".format(accessor_id), "PUT")

        return dict(changed=True, operation="clone", **cloned)


def is_policies_param_valid(policies):
    allowed_keys = set(("id", "name"))
    for link in policies:
        link_keys = set(link.keys())
        if len(allowed_keys & link_keys) == 0:
            return False
        if len(link_keys - allowed_keys) > 0:
            return False

    return True


def has_token_changed(current, updated):
    for k, v in updated.items():
        # Ignore empty secret_id
        if v is None and k.lower() == "secret_id".lower():
            continue
        if k == "policies":
            continue
        elif current.get(k) != v:
            return True

    # Compare policy links
    # TODO: Handle policies changing names
    current_names = set()
    current_ids = set()
    difference = 0
    for link in current["policies"]:
        current_names.add(link["name"])
        current_ids.add(link["id"])
        difference += 1

    for link in updated["policies"]:
        if link.get("id") in current_ids:
            difference -= 1
        elif link.get("name") in current_names:
            difference -= 1
        else:
            return True

    return difference != 0


def main():
    arg_spec = dict(
        accessor_id=dict(type="str", default=None),
        policies=dict(type="list", elements="dict", default=None),
        description=dict(type="str", default=""),
        local=dict(type="bool", default=False),
        match_description=dict(type="bool", default=False),
        state=dict(
            type="str", choices=["present", "absent", "cloned"], default="present"
        ),
        url=dict(type="str", default=os.environ.get("CONSUL_HTTP_ADDR")),
        token=dict(type="str", default=os.environ.get("CONSUL_HTTP_TOKEN")),
        validate_certs=dict(type="bool", default=True),
        client_cert=dict(type="path", default=os.environ.get("CONSUL_CLIENT_CERT")),
        client_key=dict(type="path", default=os.environ.get("CONSUL_CLIENT_KEY")),
        secret_id=dict(type="str", default=None),
    )

    module = AnsibleModule(argument_spec=arg_spec, supports_check_mode=False)

    accessor_id = module.params["accessor_id"]
    policies = module.params["policies"] or []
    description = module.params["description"]
    local = module.params["local"]
    match_description = module.params["match_description"]
    state = module.params["state"]
    url = module.params["url"]
    token = module.params["token"]
    secret_id = module.params["secret_id"]

    if not url:
        module.fail_json(msg="url must be set")
    if not token:
        module.fail_json(msg="token must be set")
    if state == "present" and not is_policies_param_valid(policies):
        module.fail_json(msg="policies must set when state == 'present'")
    if state == "cloned" and not accessor_id:
        module.fail_json(msg="accessor_id must be set when state == 'cloned'")
    if match_description and not description:
        module.fail_json(msg="description cannot be empty when matching by description")

    api = ConsulApi(module)
    consul_acl = ConsulAclToken(module, api)

    if match_description and not accessor_id:
        current_token = consul_acl.find_token(description, policies)
    elif accessor_id:
        current_token = consul_acl.get_token(accessor_id)
    else:
        current_token = None

    kwargs = dict(policies=policies, description=description, local=local, secret_id=secret_id)
    if state == "present" and current_token:
        result = consul_acl.update_token(current_token, **kwargs)
    elif state == "present":
        result = consul_acl.create_token(accessor_id=accessor_id, **kwargs)
    elif state == "cloned":
        result = consul_acl.clone_token(accessor_id)
    elif state == "absent" and accessor_id:
        result = consul_acl.delete_token(accessor_id)
    else:
        result = dict(changed=False, operation="none")

    module.exit_json(**result)


if __name__ == "__main__":
    main()
